File: gui/menu/print_graph/graph_tables.cpp
#include "graph_tables.h"


MenuGraphTables::MenuGraphTables(QWidget* parent) : 
    IodeSettingsDialog(parent)
{
    setupUi(this);

    completer = new IodeCompleter(false, false, TABLES, textEdit_table_names);
    textEdit_table_names->setCompleter(completer);

    QList<QString> q_langs;
    for(const std::string& lang: v_table_langs) q_langs << QString::fromStdString(lang);

    wTableNames = new WrapperQPlainTextEdit(label_table_names->text(), *textEdit_table_names, REQUIRED_FIELD);
    wSample = new WrapperQTextEdit(label_sample->text(), *textEdit_sample, REQUIRED_FIELD);
    wFile2 = new WrapperFileChooser(label_file_2->text(), *fileChooser_file_2, OPTIONAL_FIELD, FILE_VARIABLES, EXISTING_FILE);
    wFile3 = new WrapperFileChooser(label_file_3->text(), *fileChooser_file_3, OPTIONAL_FIELD, FILE_VARIABLES, EXISTING_FILE);
    wFile4 = new WrapperFileChooser(label_file_4->text(), *fileChooser_file_4, OPTIONAL_FIELD, FILE_VARIABLES, EXISTING_FILE);
    wFile5 = new WrapperFileChooser(label_file_5->text(), *fileChooser_file_5, OPTIONAL_FIELD, FILE_VARIABLES, EXISTING_FILE);
    wLanguage = new WrapperComboBox(label_language->text(), *comboBox_language, REQUIRED_FIELD, q_langs);
    
    mapFields["TableNames"] = wTableNames;
    mapFields["Sample"] = wSample;
    mapFields["File2"] = wFile2;
    mapFields["File3"] = wFile3;
    mapFields["File4"] = wFile4;
    mapFields["File5"] = wFile5;
    mapFields["Language"] = wLanguage;

    className = "MENU_GRAPH_TABLES";
    loadSettings();
}

MenuGraphTables::~MenuGraphTables()
{
    delete wTableNames;
    delete wSample;
    delete wFile2;
    delete wFile3;
    delete wFile4;
    delete wFile5;
    delete wLanguage;

    delete completer;
}

void MenuGraphTables::display()
{
    try
    {
        std::string pattern = wTableNames->extractAndVerify().toStdString();
        std::vector<std::string> list_names = Tables.get_names(pattern);
        QStringList qTablesList;
        for(const std::string& name : list_names) 
            qTablesList << QString::fromStdString(name);

        std::string gsample = wSample->extractAndVerify().toStdString();

        std::string file_2 = wFile2->extractAndVerify().toStdString();
        std::string file_3 = wFile3->extractAndVerify().toStdString();
        std::string file_4 = wFile4->extractAndVerify().toStdString();
        std::string file_5 = wFile5->extractAndVerify().toStdString();

        TableLang lang = TableLang (wLanguage->extractAndVerify());

        if(!file_2.empty()) load_reference_kdb(2, VARIABLES, file_2);
        if(!file_3.empty()) load_reference_kdb(3, VARIABLES, file_3);
        if(!file_4.empty()) load_reference_kdb(4, VARIABLES, file_4);
        if(!file_5.empty()) load_reference_kdb(5, VARIABLES, file_5);

        Table* table_obj;
        MainWindowAbstract* main_window = static_cast<MainWindowAbstract*>(get_main_window_ptr());
        foreach(const QString& tableName, qTablesList)
        {
            table_obj = Tables.get(tableName.toStdString());
            ComputedTableGraph* gSampleGraph = new ComputedTableGraph(table_obj, gsample);
            PlotTableDialog* plotDialog = new PlotTableDialog(gSampleGraph);
            main_window->appendPlot(plotDialog);
            delete table_obj;
        }
    }
    catch (const std::exception& e)
    {
        QMessageBox::warning(this, "WARNING", "" + QString::fromStdString(e.what()));
        return;
    }
}

void MenuGraphTables::freeReferenceKDBs()
{
    clear_all_reference_kdbs();
}

void MenuGraphTables::apply()
{
    try
    {
        QMessageBox::warning(this, "WARNING", "Apply is not yet implemented");
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
        return;
    }
}

void MenuGraphTables::setup()
{
    try
    {
        MenuFileSettings dialog(this);
        dialog.showPrintTab();
        dialog.exec();
    }
    catch (const std::exception& e)
    {
        QMessageBox::critical(this, tr("ERROR"), tr(e.what()));
        return;
    }
}

void MenuGraphTables::help()
{
	QDesktopServices::openUrl(url_manual);
}

File: gui/iode_gui/menu/print_graph/graph_tables.py
from PySide6.QtCore import Slot, Signal
from PySide6.QtWidgets import QWidget, QMessageBox, QDialog

from iode_gui.settings import MixinSettingsDialog
from iode_gui.text_edit.completer import IodeCompleter
from iode_gui.plot.plot_table import PlotTableDialog
from iode_gui.menu.file.file_settings import MenuFileSettings
from .ui_graph_tables import Ui_MenuGraphTables

from typing import List
from iode import (IodeType, TableLang, IodeFileType, Table, tables, reset_extra_files)


class MenuGraphTables(MixinSettingsDialog):
    new_plot = Signal(QDialog)

    def __init__(self, parent: QWidget=None):
        super().__init__(parent)
        self.ui = Ui_MenuGraphTables()
        self.ui.setupUi(self)
        self.prepare_settings(self.ui)

        self.ui.textEdit_table_names.setup_completer(iode_types=IodeType.TABLES)

        self.v_table_langs = list(TableLang)
        v_table_lang_names = [iode_type.name.title() for iode_type in self.v_table_langs]
        self.ui.comboBox_language.addItems(v_table_lang_names)
        self.ui.comboBox_language.setCurrentIndex(0)

        self.ui.fileChooser_file_2.enum_file_type = IodeFileType.FILE_VARIABLES
        self.ui.fileChooser_file_3.enum_file_type = IodeFileType.FILE_VARIABLES
        self.ui.fileChooser_file_4.enum_file_type = IodeFileType.FILE_VARIABLES
        self.ui.fileChooser_file_5.enum_file_type = IodeFileType.FILE_VARIABLES

        self.load_settings()

    def _free_extra_files(self):
        reset_extra_files()

    @Slot()
    def display(self):
        try:
            i_table_lang = self.ui.comboBox_language.currentIndex()
            table_lang: TableLang = self.v_table_langs[i_table_lang]
            pattern_table_names: str = self.ui.textEdit_table_names.toPlainText().strip()
            generalized_sample: str = self.ui.textEdit_sample.toPlainText().strip()
            extra_files: List[str] = []
            if self.ui.fileChooser_file_2.filepath:
                extra_files += [self.ui.fileChooser_file_2.filepath]
            if self.ui.fileChooser_file_3.filepath:
                extra_files += [self.ui.fileChooser_file_3.filepath]
            if self.ui.fileChooser_file_4.filepath:
                extra_files += [self.ui.fileChooser_file_4.filepath]
            if self.ui.fileChooser_file_5.filepath:
                extra_files += [self.ui.fileChooser_file_5.filepath]

            table_names: List[str] = tables.get_names(pattern_table_names)
            for table_name in table_names:
                table_obj: Table = tables[table_name]
                computed_table = table_obj.compute(generalized_sample, extra_files, nb_decimals=8)
                plot_dialog = PlotTableDialog(computed_table)
                self.new_plot.emit(plot_dialog)

        except Exception as e:
            QMessageBox.warning(self, "WARNING", f"Failed to plot tables '{pattern_table_names}':\n" + str(e))

    @Slot()
    def apply(self):
        QMessageBox.warning(self, "WARNING", "Apply is not yet implemented");

    @Slot()
    def setup(self):
        try:       
            dialog = MenuFileSettings(self)
            dialog.show_print_tab()
            dialog.exec()
        except Exception as e:
            QMessageBox.warning(self, "WARNING", str(e))
